### PR TITLE
Error on S-interpolator usage for assert, assume and printf

### PR DESCRIPTION
**Pull request by adkian-sifive**
This is to be backported to 3.5.x as a warning instead of an error

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

 - updated feature/API      

#### API Impact

Current `assert`, `assume` and `printf` APIs allow the Scala s-interpolator for formatting messages. This often leads to user errors where the Scala object is printed instead of the Chisel object, which leads to the messages being generally unhelpful.

After this update, using s-interpolators will throw an error about deprecation.

#### Backend Code Generation Impact

N/A
#### Desired Merge Strategy

  - Squash: The PR will be squashed and merged (choose this if you have no preference.

#### Release Notes

Using Scala s-interpolators now throws an error recommending a shift to the cf interpolator.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: src/test/scala/chiselTests/Vec.scala [hunk 1]
```diff
@@ -235,7 +235,7 @@ class IterateTester(start: Int, len: Int)(f: UInt => UInt) extends BasicTester {
   val testVec = VecInit.iterate(start.U, len)(f)
   assert(
     controlVec.asUInt() === testVec.asUInt(),
-    s"Expected Vec to be filled like $controlVec, instead creaeted $testVec\n"
+    cf"Expected Vec to be filled like $controlVec, instead creaeted $testVec\n"
```
Comment (mwachs5) [inline src/test/scala/chiselTests/Vec.scala:238]:
```suggestion
    cf"Expected Vec to be filled like $controlVec, instead created $testVec\n"
```

pull a weed

File: src/test/scala/chiselTests/VecLiteralSpec.scala [hunk 2]
```diff
@@ -205,7 +205,7 @@ class VecLiteralSpec extends ChiselFreeSpec with Utils {
 
     assertTesterPasses {
       new BasicTester {
-        chisel3.assert(outsideVecLit(0) === 0xdd.U, s"v(0)")
+        chisel3.assert(outsideVecLit(0) === 0xdd.U, cf"v(0)")
```
Comment (mwachs5) [inline src/test/scala/chiselTests/VecLiteralSpec.scala:208]: is this supposed to be ${v(0)}???

File: src/test/scala/chiselTests/VecLiteralSpec.scala [hunk 3]
```diff
@@ -216,7 +216,7 @@ class VecLiteralSpec extends ChiselFreeSpec with Utils {
...
```
Comment (mwachs5) [inline src/test/scala/chiselTests/VecLiteralSpec.scala:219]: again is this supposed to be ${v(0)}

**Comment by mwachs5** [comment]
I think this requires some fixes to the printable docs? Or can we say something there to show people the warning/error they will get if they try to printf(s"...")?

File: src/test/scala/chiselTests/VecLiteralSpec.scala (hunk 2 above)
Comment (mwachs5) [reply src/test/scala/chiselTests/VecLiteralSpec.scala:208]: or why do we need any interpolator at all?

File: core/src/main/scala/chisel3/Printf.scala [hunk 4]
```diff
@@ -100,6 +121,14 @@ object printf {
     printfId
   }
 
+  def _applyWithReset(pable: Printable)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Printf = {
+    var printfId: Printf = null
+    when(!Module.reset.asBool) {
+      printfId = printfWithoutReset(pable)
+    }
+    printfId
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/Printf.scala:129]: Why is this here? This feels like an unrelated change.

File: core/src/main/scala/chisel3/VerificationStatement.scala [hunk 5]
```diff
@@ -43,12 +43,12 @@ object assert extends VerifPrintMacrosDoc {
   // Macros currently can't take default arguments, so we need two functions to emulate defaults.
   def apply(
     cond:    Bool,
-    message: String,
+    message: => String,
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/VerificationStatement.scala:46]: In the Pull Request description you said:

> This is to be backported to 3.5.x as an error instead of a warning

I think you got that backwards that it should be backported as a warning, not an error.

We have a problem though, this change to call by-name (which is probably required for the macro to work, but I'm not 100% sure) breaks binary compatibility so cannot be backported.

Comment (adkian-sifive) [reply core/src/main/scala/chisel3/VerificationStatement.scala]: yeah makes sense, i suspected i did it the other way round. fixed

Comment (adkian-sifive) [reply core/src/main/scala/chisel3/VerificationStatement.scala]: i think the call by name change was needed by something that i've since deleted. i have reverted it

Comment (jackkoenig) [reply core/src/main/scala/chisel3/VerificationStatement.scala]: > yeah makes sense, i suspected i did it the other way round. fixed

The first line of the PR message still has the same phrasing.

File: core/src/main/scala/chisel3/Printf.scala (hunk 4 above)
Comment (adkian-sifive) [reply core/src/main/scala/chisel3/Printf.scala]: I've moved this around so that this logic is part of the function `printfWIthReset` so that i can call it from the macro

File: docs/src/explanations/printing.md [hunk 6]
```diff
@@ -103,6 +103,8 @@ Message:
 
 Notice the use of `+` between `cf` interpolated "strings". The results of `cf` interpolation can be concatenated by using the `+` operator.
 
+> :warning: The Scala s-interpolator is deprecated since 3.5 and will output the warning: "s-interpolators for Chisel assert, assume and printf statements are deprecated (since 3.5); use p or cf interpolators instead"
```
Comment (jackkoenig) [inline docs/src/explanations/printing.md:106]: Can you do an actual execution driven example rather than just a text warning?

Since it fails in this PR, you can do the following:
````
```scala mdoc:fail
class MyModule extends Module {
  val in = IO(Input(UInt(8.W)))
  printf(s"in = $in\n")
}
```
````
This should fail to compile and render with the exact error when you run `sbt docs/mdoc`

**Review by jackkoenig — APPROVED** [review]

**Comment by jackkoenig** [comment]
This is not marked with Milestone `3.5.x` because we do not want it to be backported and automerged, instead after merging we will do a manual backport and change the behavior to be just a warning.

File: docs/src/explanations/printing.md [hunk 7]
```diff
@@ -13,11 +13,23 @@ Chisel provides the `printf` function for debugging purposes. It comes in two fl
...
 ```
+
+```scala mdoc:fail
+class MyModule extends Module {
+  val in = IO(Input(UInt(8.W)))
+  printf(s"in = $in\n")
+}
+```
+
+Instead, use Chisel's `cf` interpolator as in the following examples:
```
Comment (mwachs5) [inline docs/src/explanations/printing.md:31]: Nice :)

**Review by mwachs5 — APPROVED** [review]
